torch/_functorch/_aot_autograd/fx_passes.py

```diff
@@ -0,0 +1,202 @@
+    node_to_idx = {n: i for i, n in enumerate(node_list)}
+    set_nodes_to_be_deleted = set()
+
+    # Step 1: Clean up to enforce `.set_(primal_X, ...)` usage.
+    # We replace:
+    # ```
+    # set_1 = aten.set_(primal_X, Y1)
+    # set_2 = aten.set_(set_1, Y2)
+    # ```
+    # to:
+    # ```
+    # set_1 = aten.set_(primal_X, Y1)
+    # set_2 = aten.set_(primal_X, Y2)
+    # ```
+    # Note that this process is iterative and will handle any number (>=2) of nested `.set_()`.
+    for i, n in enumerate(node_list):
+        if n.op == "call_function" and n.target is torch.ops.aten.set_.source_Tensor:
+            if (
+                n.args[0].target is torch.ops.aten.set_.source_Tensor
+                and n.args[0].args[0] in primal_inputs
+            ):
+                n.args = (n.args[0].args[0], n.args[1])
+            assert n.args[0] in primal_inputs, (
+                "Violated assumption: every `.set_` node should be setting into the primal input. "
+                f"Please report a bug to PyTorch. Violating graph: {graph}"
+            )
+
+    # Step 2: Re-functionalizing `.set_(primal_X, ...)`.
+    # Replacement conditions:
+    # 1. `.set_(primal_X, Y)` exists in graph.
+    # 2. `.set_(Y, ...)` is not called between primal_X's this `.set_(primal_X, Y)` and its next `.set_(primal_X, ...)` (or end of graph)  # noqa: B950
+    #    - This ensures primal_X and Y are semantically always meant to be the same tensor within this section of the graph.
+    # If the above two conditions are met, then within this section of the graph we will replace
+    # usage of output of this `.set_(primal_X, Y)` node with Y, and delete this `.set_(primal_X, Y)` node.
+    # For any primal input, if we have deleted the last `.set_(primal_X, ...)`, we will re-insert
+    # `.set_(primal_X, Y_last)` at the end of the graph.
+    for i, n in enumerate(node_list):
+        # For aten.set_(X, Y), X must be primal input of graph.
+        if (
+            n.op == "call_function"
+            and n.target is torch.ops.aten.set_.source_Tensor
+            and n.args[0] in primal_inputs
+        ):
+            primal_input_to_set_idx[n.args[0]].append(i)
+    for primal_input, set_idx_list in primal_input_to_set_idx.items():
+        for i in range(len(set_idx_list)):
+            set_node_idx = set_idx_list[i]
+            if i < len(set_idx_list) - 1:
+                next_set_node_idx = set_idx_list[i + 1]
+            else:
+                # The last section always ends with the graph output node.
+                next_set_node_idx = len(node_list) - 1
+            set_node = node_list[set_node_idx]
+            Y_input = set_node.args[1]
+            # Between primal_X's this `.set_(primal_X, Y)` and its next `.set_(primal_X, ...)` (or end of graph),
+            # if `.set_(Y, ...)` is not called, we will replace usage of output of this `.set_(primal_X, Y)` node with Y,
+            # and delete the `.set_(primal_X, Y)` node.
```

Comment:
hmm... one thing i'm a bit confused about is that you added the assertion on line 98 above that every `set_()` must mutate a primal, and not some other value `Y`. Does that mean:

(1) that assert will fail today on the `.data=` test you mentioned above, so we need to decide if that test still needs to be supported?
(2) keeping the assert means we no longer need to handle the case mentioned here?

Reply:
hmm so is the case here something like the below?
```
def f(primal1, primal2):
    tmp = ...
    primals1_orig = primal1.view(primal1.shape)
    primals1.set_(primals2)
    ...
    primals2.set_(tmp)
    ...
    primals1.set_(primals1_orig)
```

If I'm reading correctly, it sounds like:
(1) this pattern is needed by fsdp
(2) we will be unable to delete the `primals1.set_(primals2)` node

Doesn't that mean that the net result will be that we failed to functionalize the first `set_()` node?

Reply:
> in current version of PR, param_aliased will be replaced with ag_buffer by this FX pass.

hmm also, can you help me understand where this happens in the pass today? I think I was expecting some logic that checks if the  storages of two graph args are the same, e.g. `StorageWeakRef(param.untyped_storage()) == StorageWeakRef(param_aliased.untyped_storage())` , so the pass would know when to also functionalize the aliases

Reply:
> Doesn't that mean that the net result will be that we failed to functionalize the first set_() node?
yeah I think to avoid complexity, we should just ban this use case (which is only from FSDP1):
```
primals1.set_(primals2)
primals2.set_(tmp)
```

Reply:
> in current version of PR, param_aliased will be replaced with ag_buffer by this FX pass.

> can you help me understand where this happens in the pass today?

ah it's much more simplistic than that - I added an example to comment to make it more clear: https://github.com/pytorch/pytorch/pull/129852/files#diff-8f33d3cf8aeefa79e1c252d8749816fc03e68919aae07922b8a759a213dafba7R110